### ISSUE TITLE
[JENKINS-56064] Matrix project plugin - add a shell script to be run in top-level executor only

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#711]
<p>It would be very convenient for me to have an option to run a shell command in the top-level executor before the matrix jobs are run.</p>

<p>I want to use it to run style checks on a pull request before compilers are started. Failing the sanity check should fail the job without wasting time compiling non-conforming code.</p>

<p>What I considered:</p>
<ul>
	<li>Running style check on all executors - suboptimal and confusing. The checker does not depend on any axis parameters.</li>
	<li>Pre-SCM plugin - not good, as I need to run the checker after the code checkout, obviously.</li>
	<li>Existing code checker plugins - they all run known checkers (like cppcheck), I want to run a custom script).</li>
	<li>Multiple jobs in parallel - doesn't address the issue of time saving. Also, I want a single PR builder to talk to Stash and to post one single result at the end to prevent confusion.</li>
	<li>Chained jobs - Stash pull request builder is not flexible enough to chain pull request builders. It expects to poll Stash and post the result in the same job.</li>
</ul>

<p>What I'm using now:</p>
<ul>
	<li>One axis has a value "StyleCheck" in addition to other, legitimate values.</li>
	<li>For the "StyleCheck" build, the combination filter selects one value for every other axis.</li>
	<li>The "StyleCheck" build is the touchstone build, i.e. it runs first and makes the job fail if it fails.</li>
	<li>The execute command has a special case for the "StyleCheck" build.</li>
</ul>

<p>Needless to say, it's quite convoluted. Having an extra box for a pre-fork script would make it trivial.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/proski.png" title="proski's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=proski">proski</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-56064" target="_blank">Matrix project plugin - add a shell script to be run in top-level executor only</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: matrix-project-plugin
<li><b>label(s)</b>: matrix-project, multi-configuration, script
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2019-02-25T21:50:32+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>It would be very convenient for me to have an option to run a shell command in the top-level executor before the matrix jobs are run.</p>

<p>I want to use it to run style checks on a pull request before compilers are started. Failing the sanity check should fail the job without wasting time compiling non-conforming code.</p>

<p>What I considered:</p>
<ul>
	<li>Running style check on all executors - suboptimal and confusing. The checker does not depend on any axis parameters.</li>
	<li>Pre-SCM plugin - not good, as I need to run the checker after the code checkout, obviously.</li>
	<li>Existing code checker plugins - they all run known checkers (like cppcheck), I want to run a custom script).</li>
	<li>Multiple jobs in parallel - doesn't address the issue of time saving. Also, I want a single PR builder to talk to Stash and to post one single result at the end to prevent confusion.</li>
	<li>Chained jobs - Stash pull request builder is not flexible enough to chain pull request builders. It expects to poll Stash and post the result in the same job.</li>
</ul>

<p>What I'm using now:</p>
<ul>
	<li>One axis has a value "StyleCheck" in addition to other, legitimate values.</li>
	<li>For the "StyleCheck" build, the combination filter selects one value for every other axis.</li>
	<li>The "StyleCheck" build is the touchstone build, i.e. it runs first and makes the job fail if it fails.</li>
	<li>The execute command has a special case for the "StyleCheck" build.</li>
</ul>

<p>Needless to say, it's quite convoluted. Having an extra box for a pre-fork script would make it trivial.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins 2.150.2, Matrix Project Plugin 1.13</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/proski.png" title="proski's avatar" /> proski:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/56064?focusedId=361488&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-361488">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-56064?focusedId=361488&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-361488">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>After learning something about Jenkins internals, I'm going to close it this issue without resolving. The top-level build is considered light-weight, it runs without an executor, so adding any non-trivial jobs should be discouraged. The right solution would be to have a pipeline. If it's not working now, it should be fixed.</p>
</pre>
</details>

</ul>
</details>

<p>After learning something about Jenkins internals, I'm going to close it this issue without resolving. The top-level build is considered light-weight, it runs without an executor, so adding any non-trivial jobs should be discouraged. The right solution would be to have a pipeline. If it's not working now, it should be fixed.</p>
